Algorithms/Divide-and-Conquer/closest_pair_of_points.cpp

```diff
@@ -0,0 +1,171 @@
+// Declaration of structure
+// for x any y coordinate of a point
+struct Point
+{
+    int x, y;
+};
+ 
+// Function to sort array of points according to x coordinate
+int compareX(const void* a, const void* b)
+{
+    Point *p1 = (Point *)a, *p2 = (Point *)b;
+    return (p1->x - p2->x);
+}
+
+// Function to sort array of points according to y coordinate
+int compareY(const void* a, const void* b)
+{
+    Point *p1 = (Point *)a, *p2 = (Point *)b;
+    return (p1->y - p2->y);
+}
+
+// Function to find the Euclidean distance between two set of points
+float dist(Point p1, Point p2)
+{
+    return sqrt((p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y));
+}
+
+// Function to find the smallest distance between two points
+float small_dist(Point P[], int n)
+{
+    float min = FLT_MAX;
+    for (int i = 0; i < n; ++i)
+    {
+        for (int j = i + 1; j < n; ++j)
+        {
+            if (dist(P[i], P[j]) < min)
+                min = dist(P[i], P[j]);
+        }
+    }
+    return min;
+}
+
+// Function to find the distance beween the closest points of strip of given size
+float stripClosest(Point strip[], int size, float d)
+{
+    float min = d;
+    for (int i = 0; i < size; ++i)
+    {
+        for (int j = i + 1; j < size && (strip[j].y - strip[i].y) < min; ++j)
+        {
+            if (dist(strip[i],strip[j]) < min)
+                min = dist(strip[i], strip[j]);
+        }
+    }
+    return min;
+}
+
+// Function that finds the smallest distance between points
+// by dividing the points into left half and right half. 
+float closestUtil(Point Px[], Point Py[], int n)
+{
+    // If number of points is less than equal to 3
+    // then directly finding the closest pair of points
+    if (n <= 3)
+        return small_dist(Px, n);
+
+    // Else we require to divide points into two halves
+    int mid = n / 2;
+    Point midPoint = Px[mid];
+    Point Pyl[mid + 1];
+    Point Pyr[n - mid - 1];
+    int li = 0, ri = 0;
+
+    // Checking which points lie in left half
+    // and which one in the right half of the midpoint
+    // and storing them accodingly
+    for (int i = 0; i < n; i++)
+    {
+        if (Py[i].x <= midPoint.x)
+            Pyl[li++] = Py[i];
+        else
+            Pyr[ri++] = Py[i];
+    }
+    // Finding smallest distance separately for
+    // left half and right half.
+    float dl = closestUtil(Px, Pyl, mid);
+    float dr = closestUtil(Px + mid, Pyr, n-mid);
+    float d = min(dl, dr);
+    Point strip[n];
+    int j = 0;
+    for (int i = 0; i < n; i++)
+    {
+        if (abs(Py[i].x - midPoint.x) < d)
+            strip[j] = Py[i], j++;
+    }
+    return min(d, stripClosest(strip, j, d));
+}
+
+// Function to store x and y coordinates into separate arrays
+// Then sorting the Px array according to x coordinate
+// and sorting the Py array according to y coordinate
+float closest(Point P[], int n)
+{
+    Point Px[n];
+    Point Py[n];
+
+    // Storing x and y coordinates in separate arrays
```

Comment:
Are you storing x and y coordinates separately or just making a copy of points array in the following loop?